File: apps/web/src/lib/deploy/snapshot.ts
// Mirror of the canonical snapshot the backend stores on every Deployment row
// (see apps/backend/internal/deployment/snapshot.go). Both sides serialise
// the same shape with sorted keys; SHA-256 over the bytes is the gating
// signal for whether a deploy is needed.
//
// Shape rules — must match backend exactly:
//   - arrays sorted by id
//   - secret values omitted (frontend never sees plaintext)
//   - timestamps as int64 milliseconds since epoch — avoids tz/precision
//     mismatches between Go's RFC3339Nano and JS Date.toISOString
//   - resources.spec is the raw shape with `routes` stripped (routes live as
//     a top-level array)

import type { Resource, Route, Variable } from "@vyft/spec";

export interface Snapshot {
  resources: SnapResource[];
  routes: SnapRoute[];
  variables: SnapVariable[];
}

interface SnapResource {
  id: string;
  name: string;
  slug: string;
  kind: string;
  spec: unknown;
  updatedAt: number;
}

interface SnapRoute {
  id: string;
  resourceId: string;
  domain: string;
  path: string;
  pathType: string;
  port: number;
  tls: boolean;
  config: unknown;
  updatedAt: number;
}

// Discriminated by `kind`:
//   owned     — variable definition (project-wide if resourceId null, else
//               resource-scoped). Carries id + value/secret + updatedAt.
//   reference — env-binding of another variable into a resource under a
//               local key. Carries resourceId + key + sourceVariableId.
// Backend mirror: snapshotVariable in snapshot.go.
type SnapVariable = SnapOwnedVariable | SnapReferenceVariable;

interface SnapOwnedVariable {
  kind: "owned";
  id: string;
  resourceId: string | null;
  key: string;
  secret: boolean;
  value: string;
  updatedAt: number;
}

interface SnapReferenceVariable {
  kind: "reference";
  resourceId: string;
  key: string;
  sourceVariableId: string;
}

export function buildSnapshot(input: {
  resources: Resource[];
  routes: Route[];
  variables: Variable[];
}): Snapshot {
  const resources: SnapResource[] = input.resources
    .map((r) => ({
      id: r.id,
      name: r.name,
      slug: r.slug,
      kind: r.config.kind,
      spec: stripRoutes(r.config.spec),
      updatedAt: toMillis(r.updatedAt),
    }))
    .sort(byId);

  const routes: SnapRoute[] = input.routes
    .map((r) => ({
      id: r.id,
      resourceId: r.resourceId,
      domain: r.domain,
      path: r.path,
      pathType: r.pathType,
      port: r.port,
      tls: r.tls,
      config: r.config ?? {},
      updatedAt: toMillis(r.updatedAt),
    }))
    .sort(byId);

  const owned: SnapOwnedVariable[] = input.variables.map((v) => ({
    kind: "owned" as const,
    id: v.id,
    resourceId: v.resourceId,
    key: v.key,
    secret: v.secret,
    value: v.secret ? "" : (v.value ?? ""),
    updatedAt: toMillis(v.updatedAt),
  }));

  // References: each Variable carries `usedBy` — resources that import it
  // under some local key. Expand to per-reference rows so adding/removing
  // an import flips the snapshot hash.
  const references: SnapReferenceVariable[] = input.variables.flatMap((v) =>
    (v.usedBy ?? []).map((u) => ({
      kind: "reference" as const,
      resourceId: u.id,
      key: u.key,
      sourceVariableId: v.id,
    })),
  );

  // Sort matches backend: owned by id; reference by (resourceId, key).
  // Kind groups separately ("owned" < "reference" lexicographically).
  owned.sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
  references.sort((a, b) => {
    if (a.resourceId !== b.resourceId)
      return a.resourceId < b.resourceId ? -1 : 1;
    return a.key < b.key ? -1 : a.key > b.key ? 1 : 0;
  });
  const variables: SnapVariable[] = [...owned, ...references];

  return { resources, routes, variables };
}

function toMillis(s: string): number {
  return new Date(s).getTime();
}

function byId(a: { id: string }, b: { id: string }) {
  return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
}

// stripRoutes removes the embedded `routes` array from a resource's spec —
// matches the raw DB column shape used by the backend snapshot.
function stripRoutes(spec: unknown): unknown {
  if (spec == null || typeof spec !== "object") return spec ?? {};
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(spec as Record<string, unknown>)) {
    if (k === "routes") continue;
    out[k] = v;
  }
  return out;
}

// canonicalStringify produces sorted-key JSON so snapshots from either side
// hash byte-identically. Recursive: arrays preserve order (already sorted at
// build time), objects key-sorted at every level.
export function canonicalStringify(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) {
    return "[" + value.map(canonicalStringify).join(",") + "]";
  }
  const keys = Object.keys(value as Record<string, unknown>).sort();
  const entries = keys.map(
    (k) =>
      JSON.stringify(k) +
      ":" +
      canonicalStringify((value as Record<string, unknown>)[k]),
  );
  return "{" + entries.join(",") + "}";
}

// stripVolatile removes fields that mutate without changing deployable content
// (currently `updatedAt`). Restore bumps the resource row's `updated` column
// even though the spec matches a prior deployment — keeping updatedAt in the
// hash would mark a restored project as "has changes" against its own
// snapshot. Recurses into arrays + objects.
function stripVolatile(value: unknown): unknown {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) return value.map(stripVolatile);
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
    if (k === "updatedAt") continue;
    out[k] = stripVolatile(v);
  }
  return out;
}

export async function snapshotHash(snapshot: unknown): Promise<string> {
  const text = canonicalStringify(stripVolatile(snapshot));
  const buf = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(text),
  );
  return Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

File: apps/backend/internal/deployment/snapshot.go
package deployment

import (
	"context"
	"encoding/json"
	"sort"

	"github.com/google/uuid"

	"github.com/vyftlabs/vyft/apps/backend/internal/db/sqlc"
	"github.com/vyftlabs/vyft/apps/backend/internal/platform/pgxid"
)

// snapshotShape is the canonical reduced JSON shape stored on every
// deployment row. The frontend reconstructs an identical shape from its
// cached query data and hashes both — match means "no changes since this
// deployment".
//
// Rules so the shape round-trips byte-identically across sides:
//   - arrays sorted by id
//   - secret values omitted (frontend can't see plaintext); the row's
//     updatedAt covers value changes via the touch_updated trigger
//   - timestamps as int64 milliseconds since epoch (UTC). Avoids the tz +
//     nano-precision mismatches that string formats have between Go and JS.
//   - resources.spec is parsed JSON (map), NOT raw bytes. Both sides
//     canonical-stringify recursively so key order doesn't matter, but the
//     contents must be parseable and equivalent.
//   - reference variables (kind="reference") live in the same Variables
//     array alongside owned definitions. Adding/removing a reference flips
//     the hash even when no owned row changed. Frontend reconstructs them
//     from Variable.usedBy (each owned carries the resources referencing it).
type snapshotShape struct {
	Resources []snapshotResource `json:"resources"`
	Routes    []snapshotRoute    `json:"routes"`
	Variables []snapshotVariable `json:"variables"`
}

type snapshotResource struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Slug      string `json:"slug"`
	Kind      string `json:"kind"`
	Spec      any    `json:"spec"`
	UpdatedAt int64  `json:"updatedAt"`
}

type snapshotRoute struct {
	ID         string `json:"id"`
	ResourceID string `json:"resourceId"`
	Domain     string `json:"domain"`
	Path       string `json:"path"`
	PathType   string `json:"pathType"`
	Port       int32  `json:"port"`
	TLS        bool   `json:"tls"`
	Config     any    `json:"config"`
	UpdatedAt  int64  `json:"updatedAt"`
}

// snapshotVariable is discriminated by Kind. Marshaled per-kind via custom
// MarshalJSON so the on-disk shape only contains fields relevant to that
// kind — keeps the hash stable with the frontend's discriminated union.
//
//	owned     — definition row. Has id, key, value/secret, updatedAt. Lives
//	            either project-wide (ResourceID nil) or resource-scoped.
//	reference — env-binding of another variable into a resource. Has
//	            resourceId, key (local name), and sourceVariableId pointing
//	            to the owned variable being referenced.
type snapshotVariable struct {
	Kind             string
	ID               string
	ResourceID       *string
	Key              string
	Secret           bool
	Value            string
	UpdatedAt        int64
	SourceVariableID string
}

func (v snapshotVariable) MarshalJSON() ([]byte, error) {
	if v.Kind == "reference" {
		return json.Marshal(struct {
			Kind             string `json:"kind"`
			ResourceID       string `json:"resourceId"`
			Key              string `json:"key"`
			SourceVariableID string `json:"sourceVariableId"`
		}{"reference", derefStr(v.ResourceID), v.Key, v.SourceVariableID})
	}
	return json.Marshal(struct {
		Kind       string  `json:"kind"`
		ID         string  `json:"id"`
		ResourceID *string `json:"resourceId"`
		Key        string  `json:"key"`
		Secret     bool    `json:"secret"`
		Value      string  `json:"value"`
		UpdatedAt  int64   `json:"updatedAt"`
	}{"owned", v.ID, v.ResourceID, v.Key, v.Secret, v.Value, v.UpdatedAt})
}

func (v *snapshotVariable) UnmarshalJSON(b []byte) error {
	var probe struct {
		Kind string `json:"kind"`
	}
	if err := json.Unmarshal(b, &probe); err != nil {
		return err
	}
	if probe.Kind == "reference" {
		var r struct {
			Kind             string `json:"kind"`
			ResourceID       string `json:"resourceId"`
			Key              string `json:"key"`
			SourceVariableID string `json:"sourceVariableId"`
		}
		if err := json.Unmarshal(b, &r); err != nil {
			return err
		}
		rid := r.ResourceID
		*v = snapshotVariable{Kind: "reference", ResourceID: &rid, Key: r.Key, SourceVariableID: r.SourceVariableID}
		return nil
	}
	var o struct {
		Kind       string  `json:"kind"`
		ID         string  `json:"id"`
		ResourceID *string `json:"resourceId"`
		Key        string  `json:"key"`
		Secret     bool    `json:"secret"`
		Value      string  `json:"value"`
		UpdatedAt  int64   `json:"updatedAt"`
	}
	if err := json.Unmarshal(b, &o); err != nil {
		return err
	}
	*v = snapshotVariable{Kind: "owned", ID: o.ID, ResourceID: o.ResourceID, Key: o.Key, Secret: o.Secret, Value: o.Value, UpdatedAt: o.UpdatedAt}
	return nil
}

// buildSnapshot loads env-scoped state and renders the canonical shape.
// Stored in the deployment row's `snapshot` JSONB column.
func (s *Service) buildSnapshot(ctx context.Context, projectID, envID uuid.UUID) ([]byte, error) {
	return buildSnapshotWith(ctx, s.db.Q, projectID, envID)
}

// buildSnapshotWith is the tx-aware variant used inside WithTx callbacks
// that need to snapshot the post-write state.
func buildSnapshotWith(ctx context.Context, q *sqlc.Queries, projectID, envID uuid.UUID) ([]byte, error) {
	pid := pgxid.PgUUID(projectID)
	eid := pgxid.PgUUID(envID)

	resources, err := q.ListResourcesByProject(ctx, pid)
	if err != nil {
		return nil, err
	}
	routes, err := q.ListRoutesByProjectEnv(ctx, sqlc.ListRoutesByProjectEnvParams{
		ProjectID:     pid,
		EnvironmentID: eid,
	})
	if err != nil {
		return nil, err
	}
	variables, err := q.ListVariablesByProjectEnv(ctx, sqlc.ListVariablesByProjectEnvParams{
		ProjectID:     pid,
		EnvironmentID: eid,
	})
	if err != nil {
		return nil, err
	}
	imports, err := q.ListResourceImportsByEnv(ctx, sqlc.ListResourceImportsByEnvParams{
		ProjectID:     pid,
		EnvironmentID: eid,
	})
	if err != nil {
		return nil, err
	}

	out := snapshotShape{
		Resources: make([]snapshotResource, 0, len(resources)),
		Routes:    make([]snapshotRoute, 0, len(routes)),
		Variables: make([]snapshotVariable, 0, len(variables)+len(imports)),
	}

	for _, r := range resources {
		spec := parseJSONOrEmpty(r.Spec)
		out.Resources = append(out.Resources, snapshotResource{
			ID:        uuid.UUID(r.ID.Bytes).String(),
			Name:      r.Name,
			Slug:      r.Slug,
			Kind:      r.Kind,
			Spec:      spec,
			UpdatedAt: r.Updated.Time.UnixMilli(),
		})
	}
	sort.Slice(out.Resources, func(i, j int) bool { return out.Resources[i].ID < out.Resources[j].ID })

	for _, rt := range routes {
		cfg := parseJSONOrEmpty(rt.Config)
		out.Routes = append(out.Routes, snapshotRoute{
			ID:         uuid.UUID(rt.ID.Bytes).String(),
			ResourceID: uuid.UUID(rt.ResourceID.Bytes).String(),
			Domain:     rt.Domain,
			Path:       rt.Path,
			PathType:   string(rt.PathType),
			Port:       rt.Port,
			TLS:        rt.Tls,
			Config:     cfg,
			UpdatedAt:  rt.Updated.Time.UnixMilli(),
		})
	}
	sort.Slice(out.Routes, func(i, j int) bool { return out.Routes[i].ID < out.Routes[j].ID })

	for _, v := range variables {
		secret := v.Secret != nil && *v.Secret
		var value string
		if !secret && v.Value != nil {
			value = *v.Value
		}
		var rid *string
		if v.ResourceID.Valid {
			s := uuid.UUID(v.ResourceID.Bytes).String()
			rid = &s
		}
		out.Variables = append(out.Variables, snapshotVariable{
			Kind:       "owned",
			ID:         uuid.UUID(v.ID.Bytes).String(),
			ResourceID: rid,
			Key:        v.Key,
			Secret:     secret,
			Value:      value,
			UpdatedAt:  v.Updated.Time.UnixMilli(),
		})
	}
	for _, imp := range imports {
		rid := uuid.UUID(imp.ResourceID.Bytes).String()
		out.Variables = append(out.Variables, snapshotVariable{
			Kind:             "reference",
			ResourceID:       &rid,
			Key:              imp.Key,
			SourceVariableID: uuid.UUID(imp.VariableID.Bytes).String(),
		})
	}
	sort.Slice(out.Variables, func(i, j int) bool {
		a, b := out.Variables[i], out.Variables[j]
		if a.Kind != b.Kind {
			return a.Kind < b.Kind
		}
		// owned: sort by id (always set); reference: sort by (resourceId, key).
		if a.Kind == "owned" {
			return a.ID < b.ID
		}
		if ridA, ridB := derefStr(a.ResourceID), derefStr(b.ResourceID); ridA != ridB {
			return ridA < ridB
		}
		return a.Key < b.Key
	})

	return json.Marshal(out)
}

func derefStr(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

// filterByResourceChanges walks the given deployments newest→oldest and
// keeps those whose snapshot slice for the resource differs from the
// next-older deployment's slice. Rows must already be ordered newest→
// oldest (as returned by List).
func filterByResourceChanges(rows []sqlc.Deployment, resourceID uuid.UUID) ([]sqlc.Deployment, error) {
	if len(rows) == 0 {
		return nil, nil
	}
	rid := resourceID.String()
	slices := make([]string, len(rows))
	for i, r := range rows {
		s, err := resourceSlice(r.Snapshot, rid)
		if err != nil {
			return nil, err
		}
		slices[i] = s
	}
	out := make([]sqlc.Deployment, 0, len(rows))
	for i, r := range rows {
		// Compare against the next-older deployment. The oldest row has
		// no predecessor — treat empty as the baseline so "service first
		// appears" still counts as a change.
		var prev string
		if i+1 < len(rows) {
			prev = slices[i+1]
		}
		if slices[i] != prev {
			out = append(out, r)
		}
	}
	return out, nil
}

// resourceSlice canonical-stringifies the subset of the snapshot that
// belongs to a single resource: its own resource entry + routes/variables
// whose resourceId matches. Returns "" for empty snapshots.
func resourceSlice(raw []byte, rid string) (string, error) {
	if len(raw) == 0 {
		return "", nil
	}
	var snap snapshotShape
	if err := json.Unmarshal(raw, &snap); err != nil {
		return "", err
	}
	out := snapshotShape{}
	for _, r := range snap.Resources {
		if r.ID == rid {
			out.Resources = append(out.Resources, r)
		}
	}
	for _, rt := range snap.Routes {
		if rt.ResourceID == rid {
			out.Routes = append(out.Routes, rt)
		}
	}
	for _, v := range snap.Variables {
		// Both kinds: owned-by-resource and reference rows are scoped via
		// ResourceID. Shared owned vars (ResourceID nil) belong to the
		// project, not any single resource — excluded from per-resource slice.
		if v.ResourceID != nil && *v.ResourceID == rid {
			out.Variables = append(out.Variables, v)
		}
	}
	// Resource absent from this deployment — match the empty-snapshot sentinel
	// so filterByResourceChanges treats "didn't exist" the same as "no rows".
	if len(out.Resources) == 0 && len(out.Routes) == 0 && len(out.Variables) == 0 {
		return "", nil
	}
	bytes, err := json.Marshal(out)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}

// parseJSONOrEmpty unmarshals raw JSON bytes into a generic value, returning
// an empty object if the input is empty/null. Storing parsed (not raw) form
// in the snapshot gives Go's encoder a chance to canonical-sort map keys.
func parseJSONOrEmpty(raw []byte) any {
	if len(raw) == 0 {
		return map[string]any{}
	}
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return map[string]any{}
	}
	if v == nil {
		return map[string]any{}
	}
	return v
}
